### DISCUSSION TITLE
Minor feedback on RC as a I go along with the upgrade

### DISCUSSION THREAD

**Discussion by pnmcosta** [pocketbase/pocketbase#5631] [Q&A]
Hi Gani,

Found a couple of minor things as I am going along with the upgrade for you and the community to be aware of:

1. Even though we're able to make changes to collection system fields in the admin, in this case I wanted to ensure email is non-empty:

![image](https://github.com/user-attachments/assets/c2a5e4c6-0617-456b-8595-be08434a80a7)

The migration generated looks as follows (redacted the collection id):

```
package migrations

import (
	"encoding/json"

	"github.com/pocketbase/pocketbase/core"
	m "github.com/pocketbase/pocketbase/migrations"
)

func init() {
	m.Register(func(app core.App) error {
		collection, err := app.FindCollectionByNameOrId("XXX")
		if err != nil {
			return err
		}

		// update field
		if err := json.Unmarshal([]byte(`[{
			"exceptDomains": null,
			"hidden": false,
			"id": "_pbf_auth_email_",
			"name": "email",
			"onlyDomains": null,
			"presentable": false,
			"required": true,
			"system": true,
			"type": "email"
		}]`), &collection.Fields); err != nil {
			return err
		}

		return app.Save(collection)
	}, func(app core.App) error {
		collection, err := app.FindCollectionByNameOrId("XXX")
		if err != nil {
			return err
		}

		// update field
		if err := json.Unmarshal([]byte(`[{
			"exceptDomains": null,
			"hidden": false,
			"id": "_pbf_auth_email_",
			"name": "email",
			"onlyDomains": null,
			"presentable": false,
			"required": false,
			"system": true,
			"type": "email"
		}]`), &collection.Fields); err != nil {
			return err
		}

		return app.Save(collection)
	})
}

```

But when it is applied I get the following error `Failed to apply migration 1728412474_updated_accounts.go: fields: System fields cannot be deleted or renamed.`. I think that if all system fields can't indeed by modified, the admin UI should not allow to do so as it could induce the user in error, only noticing when the app restarts that the migration will fail (if using migrations of course).

2. The `model.RefreshId()` method was removed, I found `core.GenerateDefaultRandomId()` function as a direct replacement to use, I think the upgrade documentation could reflect this.

If you could keep this thread open, I'll update with anything I consider minor going forth.

**Comment by ganigeorgiev** [comment]
The "system field error" is most likely because the above will replace all other collection fields with only a single `email` and not because of the `email` field change.

The autogenerated Go template migration is incorrect and need to be fixed. I'll look at it tomorrow.

Regarding `RefreshId()`, you can use `core.GenerateDefaultRandomId()` if you want but note that the generated id is random and there is no guarantee that it will comply with the field `Pattern` option.
Instead you can use the `:autogenerate` text field modifier like `record.Set("id:autogenerate", "")`.
All modifiers will be documented with the final release but you can refer to the godoc of the fields for now - https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.0-rc3/core#TextField.

**Reply by pnmcosta** [reply]
Oh I see that now. Many thanks.

**Reply by pnmcosta** [reply]
Running the latest `develop` branch, just here to report that 1. is no longer an issue/concern, it's totally fixed.

Many thanks Gani.

**Comment by pnmcosta** [comment]
OMG! I just came across [core.BaseRecordProxy](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.0-rc3/core#BaseRecordProxy) this is amazing and needs further documentation 😅 

Here's a snippet:

```
type Account struct {
	core.BaseRecordProxy
}

record := core.NewRecord(collection)
record.Set("id:autogenerate", "")

account := &Account{}
account.SetProxyRecord(record)

// you then have access to all core.Record fields and methods 🥳
// e.g.

func (r Account) IsActive() bool {
	return r.GetDateTime("deleted").IsZero() 
}
```

Outstanding work Gani 🙇 this is heaps better than the previous version custom struct types

**Comment by pnmcosta** [comment]
@ganigeorgiev think that collection snapshots taken before latest rc8 should be regenerated, I'm getting this after upgrade to latest:

```
Failed to apply migration 1728379340_collections_snapshot.go: collections: Data validations failed for collection "_mfas" (_pbc_1285377323):
        {
          "deleteRule": "System collection API rule cannot be changed."
        }.
```

Think the cause is 

> Lock the _otps and _mfas system collections Delete API rule for superusers only.

from https://github.com/pocketbase/pocketbase/blob/develop/CHANGELOG.md

**Reply by ganigeorgiev** [reply]
Yes, the error is caused by the recent change. I've updated the release notes to mention it.

**Comment by pnmcosta** [comment]
@ganigeorgiev I wonder if it's worth to [instance](https://github.com/pocketbase/pocketbase/blob/develop/apis/record_crud.go#L197) record forms when [triggering](https://github.com/pocketbase/pocketbase/blob/develop/apis/record_crud.go#L211) instead so that its save operations carry the request context instead of new background one? https://github.com/pocketbase/pocketbase/blob/develop/forms/record_upsert.go#L39

Reason being that I have a couple of use cases, where `OnRecordUpdateRequest` I would like to set a context key from a endpoint's custom querystring/form data field (not on collection fields) that down chain `OnRecordUpdate` or `OnRecordAfterUpdateSuccess` I could act/bypass on.

I have considered custom record data but I'm not sure if the record pointer remains across all events, and I also have another use case where I would like to persist a struct as a context value that should be lazy loaded across events as needed and being able to leverage a request context in record events seems like a good feature to have, as any record `Save` could leverage similar logic.

**Reply by ganigeorgiev** [reply]
I don't know your use case but I'm tired people constantly pinging me for no reason and complaining. Please just use something else and move on.

**Reply by pnmcosta** [reply]
I'm sorry if I came across as complaining it was unintended. Quite the contrary, I don't intend to move on I think this latest release is a step up IMO! I have explained my use cases and suggestion, if not interested/suitable/feasible that's fine, but I'm also happy to elaborate further if wanted.
